dbos/workflow.go

```diff
@@ -738,37 +747,106 @@ func RunAsStep[P any, R any](ctx DBOSContext, fn GenericStepFunc[P, R], input P)
 		return *new(R), newStepExecutionError("", "", "step function cannot be nil")
 	}
 
-	// Type-erase the function based on its actual type
-	typeErasedFn := StepFunc(func(ctx context.Context, input any) (any, error) {
-		typedInput, ok := input.(P)
-		if !ok {
-			return nil, newStepExecutionError("", "", fmt.Sprintf("unexpected input type: expected %T, got %T", *new(P), input))
-		}
-		return fn(ctx, typedInput)
-	})
+	// Check if fn is a valid function
+	fnType := reflect.TypeOf(fn)
+	if fnType.Kind() != reflect.Func {
+		return *new(R), fmt.Errorf("fn must be a function, got %T", fn)
+	}
 
-	typeErasedStepNameToStepName[runtime.FuncForPC(reflect.ValueOf(typeErasedFn).Pointer()).Name()] = runtime.FuncForPC(reflect.ValueOf(fn).Pointer()).Name()
+	// Check that the first argument implements context.Context
+	if fnType.NumIn() == 0 || fnType.In(0) != reflect.TypeOf((*context.Context)(nil)).Elem() {
```

Comment:
This is statically guaranteed, no?

Reply:
No the function is of type `any`, so we have no static guarantees about what the first argument here.

I can simplify these checks to wrap everything under the "safe execution" logic, which recovers from panics. Not exactly sure what the error message would be. (Tho I'd be inclined to simplify this code as much as possible.)